File: src/offer30.MinInStack/min_in_stack.cpp
class Solution1 {
public:
    stack<int> m_data;
    stack<int> m_min;
    void push(int value) {
        m_data.push(value);
        if(m_min.size() == 0 || value < m_min.top()){
            m_min.push(value);
        }
        else{
            m_min.push(m_min.top());
        }
    }
    void pop() {
        m_data.pop();
        m_min.pop();
    }
    int top() {
        return m_data.top();
    }
    int min() {
        return m_min.top();
    }
};

class Solution2 {
public:
    int min_data = INT_MAX;
    stack<int> sta;
    void push(int value) {
        if (value < min_data){
            sta.push(min_data);
            min_data = value;
        }
        sta.push(value);
    }
    void pop() {
        if (sta.top() == min_data){
            sta.pop();
            min_data = sta.top();
            sta.pop();
        }
        else{
            sta.pop();
        }
    }
    int top() {
        return sta.top();
    }
    int min() {
        return min_data;
    }
};

class Solution3 {
public:
    class Node{
    public:
        int value;
        int min;
        Node* next;
        Node(int x, int min):value(x), min(min),next(nullptr) {}
    };
    Node* head;
    
    void push(int value) {
        if (head == nullptr){
            head = new Node(value, value);
        }
        else{
            Node* node;
            if (value < head->min){
                node = new Node(value, value);
            }
            else{
                node = new Node(value, head->value);
            }
            node->next = head;
            head = node;
        }
    }
    void pop() {
        Node* del = head;
        head = head->next;
        delete del;
    }
    int top() {
        return head->value;
    }
    int min() {
        return head->min;
    }
};

File: src/offer30.MinInStack/min_in_stack.java
//1
import java.util.Stack;

public class Solution {
    Stack<Integer> numStack = new Stack<>();
    Stack<Integer> minStack = new Stack<>();
    public void push(int node) {
        numStack.push(node);
        if(minStack.size() == 0 || node < minStack.peek()){
            minStack.push(node);
        }else{
            minStack.push(minStack.peek());
        }
    }
    
    public void pop() {
        minStack.pop();
        numStack.pop();
    }
    
    public int top() {
        return numStack.peek();
    }
    
    public int min() {
        return minStack.peek();
    }
}
//2
import java.util.Stack;

public class Solution {
    Stack<Integer> stack = new Stack<>();
    int min = Integer.MAX_VALUE;
    public void push(int node) {
        if(node < min){
            stack.push(min);
            min = node;
        }
        stack.push(node);
    }
    
    public void pop() {
        int cur = stack.pop();
        if(cur == min){
            min = stack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int min() {
        return min;
    }
}
//3
import java.util.Stack;

public class Solution {

    public class Node{
        int val;
        int min;
        Node next;
        public Node(int val, int min){
            this.val = val;
            this.min = min;
        }
    }
    Node head;
    public void push(int node) {
        if(head == null){
            head = new Node(node, node);
        }else{
            Node cur;
            if(node < head.min){
                cur = new Node(node, node);
            }else{
                cur = new Node(node, head.min);
            }
            cur.next = head;
            head = cur;
        }
    }
    
    public void pop() {
        head = head.next;
    }
    
    public int top() {
        return head.val;
    }
    
    public int min() {
        return head.min;
    }
}